### ISSUE TITLE
Can FlightRecorder set the stream key of collective and P2P OP into the entry.

### ISSUE THREAD

**Issue by workingloong** [pytorch/pytorch#137974]
### 🚀 The feature, motivation and pitch

Flight recorder can dump uncompleted NCCL OPs from the trace buffer if the distributed training hangs.  In codes of ProcessGroupNCCL.cpp, the collective OPs and batch P2P OPs are launched on the same stream. The single P2P are launched on another stream. We can select the first stuck OP on each stream if the entry of FlightRecorder has the stream key. We only need to analyze the first stuck OP on each stream to diagnose the hang training because other OPs on the same stream are blocked by the first OP.
https://github.com/pytorch/pytorch/blob/b7be4b1e4803351d05d5ed5b219532f7dc5daaa7/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L3161-L3170

**Comment by wconstab** [comment]
I think we could add the stream info into the flight buffer.  Otoh I'm not convinced we need to: we can also just infer the stream based on the type of op.  Collectives all share a stream, and p2p ops are going to use individual streams keyed by "src:dst" ranks.  So we can do the same analysis and find the first stuck op per stream inside the resulting data anyway.  What do you think?

cc @c-p-i-o  @fduwjj @kwen2501

**Comment by fduwjj** [comment]
@wconstab if we still keep the assumption that each sub-PG maps to a separate stream. Then, yes, we can infer that without logging stream info.

**Comment by workingloong** [comment]
> ption that each sub-PG maps to a separate strea

We cannot infer the stream of p2p op because batch p2p ops (batch_isend_irecv) shares the stream with the Collectives. Now, we cannot distinguish the p2p op is a batch p2p or a single p2p by the current info. Because the isP2P=true for both batch p2p and single p2p.

**Comment by wconstab** [comment]
Oh, if we can't distinguish between batch p2p and individual p2p that's something we should fix.

**Comment by workingloong** [comment]
> Oh, if we can't distinguish between batch p2p and individual p2p that's something we should fix.

A simple solution that we can set `isP2P=false` for batch p2p. But this may be confused and not readable.

**Comment by wconstab** [comment]
It's also simple to add another flag. I'll let @c-p-i-o weigh in, he was updating this code recently and may have thought about this issue.

**Comment by fduwjj** [comment]
@workingloong for `batch_isend_irecv` we call start_coalesce and end_coalesce so we don't bump up seqP2P_ for the send/recv happening in between, would this make it different from P2P op?

**Comment by workingloong** [comment]
> @workingloong for `batch_isend_irecv` we call start_coalesce and end_coalesce so we don't bump up seqP2P_ for the send/recv happening in between, would this make it different from P2P op?

It is difficult to distinguish by the `seqP2P_`. For example, there are 3 ops, the first is a single P2P and others are `batch_isend_irecv`. The `seqP2P_` of `batch_isend_irecv` will be same as the first single P2P. We can make it different if we set `seqP2P_=0` for `batch_isend_irecv` and set `seqCollective_=0` for the single P2P.

**Comment by workingloong** [comment]
Otoh it dose not make sense to set `seqCollective_` for P2P ops and `seqP2P_` for collectives.

**Comment by fduwjj** [comment]
@workingloong 

The corner case is indeed valid... 

Also when you say

> Otoh it dose not make sense to set seqCollective_ for P2P ops and seqP2P_ for collectives.

I think we don't do that right? it's just these two private variables are all in ProcessGroupNCCL, we record them to FR anyway. But I guess we can make `seqCollective_` only logged for non-P2P collectives? cc: @c-p-i-o

**Comment by workingloong** [comment]
@fduwjj Yeah. But, it is easy and simple to add a flag `stream_key` into the entry of FlightRecorder.
